src/app/services/xml-parsers/witnesses-parser.service.ts

```diff
@@ -0,0 +1,123 @@
+    private editionDataService: EditionDataService,
+    private genericParserService: GenericParserService,
+  ) {
+  }
+
+  private parseLists(document: XMLElement): WitnessesList {
+    const parsedList: WitnessesList = {
+      witnesses: {},
+      originalGroup: {},
+      indexes: {
+        witnesses: [],
+        groups: [],
+        encodingStructure: [],
+      },
+    };
+    Array.from(document.querySelectorAll<XMLElement>(this.tagNamesMap.witListTagName))
+      .map((list) => {
+        !isNestedInElem(list, list.tagName) ? this.parseWitnesses(list, parsedList) : this.parseWitnessesGroups(list, parsedList);
+      });
+
+    return parsedList;
+  }
+
+  private parseWitnesses(list: XMLElement, parsedList: WitnessesList) {
+    Array.from(list.querySelectorAll<XMLElement>(this.tagNamesMap.witTagName))
+      .map((wit) => {
+        const parsedWit = this.parseWitness(wit);
+        parsedList.witnesses[parsedWit.id] = parsedWit;
+        parsedList.indexes.witnesses.push(wit.getAttribute('xml:id'));
+        if (parsedWit.group === undefined) {
+          parsedList.indexes.encodingStructure.push(wit.getAttribute('xml:id'));
+        }
+        if (parsedWit.group && !parsedList.indexes.encodingStructure.includes(parsedWit.group)) {
+          parsedList.indexes.encodingStructure.push(parsedWit.group);
+        }
+      });
+  }
+
+  private parseWitness(wit: XMLElement): Witness {
+    return {
+      id: wit.getAttribute('xml:id') || xpath(wit),
+      attributes: this.genericParserService.parseAttributes(wit),
+      content: this.parseWitnessContent(wit),
+      group: this.parseParentGroupId(wit),
+    };
+  }
+
+  private parseWitnessContent(wit: XMLElement): Description {
+    const content = [];
+    Array.from(wit.childNodes)
+      .map((child: XMLElement) => content.push(this.genericParserService.parse(child)));
+
+    return content;
+  }
+
+  private parseWitnessesGroups(list: XMLElement, parsedList: WitnessesList) {
+    const parsedGroup = this.parseWitnessGroup(list);
+    parsedList.originalGroup[parsedGroup.id] = parsedGroup;
+    parsedList.indexes.groups.push(parsedGroup.id);
+  }
+
+  private parseWitnessGroup(list: XMLElement): WitnessGroup {
+    return {
+      id: list.getAttribute('xml:id') || xpath(list),
+      name: this.parseGroupName(list) || replaceNotWordChar(list.getAttribute('xml:id')) || xpath(list),
+      attributes: this.genericParserService.parseAttributes(list),
+      content: this.parseGroupContent(list),
+      group: this.parseParentGroupId(list) || undefined,
+    };
+  }
+
+  private parseGroupName(list: XMLElement) {
+    const groupTagName = 'head';
+    const groupEl = Array.from(list.children)
+      .find((child) => child.nodeName === groupTagName);
+
+    return groupEl ? groupEl.textContent : undefined;
+  }
+
+  private parseGroupContent(list: XMLElement): Description {
+    const content = [];
+    Array.from(list.children)
+      .filter(() => Object.keys(this.tagNamesMap))
```

Comment:
Your intentions are unclear

Reply:
I apply a filter that retrieve only the elements in the `tagNamesMap` (`listWit` and `witness`) to get the content of a specific group, that is which witnesses or groups belong to the current group. How can I make this code clearer?

Reply:
You are actually filtering by empty vs non empty `this.tagNamesMap` keys. 
If that is what you want, make it explicit `Object.keys(this.tagNamesMap).lenght > 0`.